ethash/src/compute.rs

```diff
@@ -269,57 +271,81 @@ pub fn light_compute(light: &Light, header_hash: &H256, nonce: u64) -> ProofOfWo
 }
 
 fn hash_compute(light: &Light, full_size: usize, header_hash: &H256, nonce: u64) -> ProofOfWork {
+	macro_rules! make_const_array {
+		($n:expr, $value:expr) => {{
+			// We use explicit lifetimes to ensure that val's borrow is invalidated until the
+			// transmuted val dies.
+			unsafe fn make_const_array<'a, T, U>(val: &'a mut [T]) -> &'a mut [U; $n] {
+				use ::std::mem;
+
+				debug_assert_eq!(val.len() * mem::size_of::<T>(), $n * mem::size_of::<U>());
+				mem::transmute(val.as_mut_ptr())
+			}
+
+			make_const_array($value)
+		}}
+	}
+
 	if full_size % MIX_WORDS != 0 {
 		panic!("Unaligned full size");
 	}
-	// pack hash and nonce together into first 40 bytes of s_mix
-	let mut s_mix: [Node; MIX_NODES + 1] = [Node::default(), Node::default(), Node::default()];
-	unsafe { ptr::copy_nonoverlapping(header_hash.as_ptr(), s_mix.get_unchecked_mut(0).bytes.as_mut_ptr(), 32) };
-	unsafe { ptr::copy_nonoverlapping(mem::transmute(&nonce), s_mix.get_unchecked_mut(0).bytes[32..].as_mut_ptr(), 8) };
 
-	// compute sha3-512 hash and replicate across mix
 	unsafe {
+		// pack hash and nonce together into first 40 bytes of s_mix
+		let mut s_mix: [Node; MIX_NODES + 1] = mem::uninitialized();
+
+		ptr::copy_nonoverlapping(
+			header_hash.as_ptr(),
+			s_mix.get_unchecked_mut(0).bytes.as_mut_ptr(),
+			32,
+		);
+		ptr::copy_nonoverlapping(
+			mem::transmute(&nonce),
+			s_mix.get_unchecked_mut(0).bytes[32..].as_mut_ptr(),
+			8
+		);
+
+		// compute sha3-512 hash and replicate across mix
 		sha3::sha3_512(s_mix.get_unchecked_mut(0).bytes.as_mut_ptr(), NODE_BYTES, s_mix.get_unchecked(0).bytes.as_ptr(), 40);
-		let (f_mix, mut mix) = s_mix.split_at_mut(1);
+		let (f_mix, mix) = s_mix.split_at_mut(1);
+		let mix_words: &mut [u32; MIX_WORDS] = make_const_array!(MIX_WORDS, mix);
+
 		for w in 0..MIX_WORDS {
-			*mix.get_unchecked_mut(0).as_words_mut().get_unchecked_mut(w) = *f_mix.get_unchecked(0).as_words().get_unchecked(w % NODE_WORDS);
+			*mix_words.get_unchecked_mut(w) =
```

Comment:
It got a pointer to `mix[1]`, transmuted it to a `[u32; NODE_WORDS]`, and then read between index 0 and `NODE_WORDS * 2`. Basically reading into `mix[2]`, yeah. Now it transmutes to a `[u32; NODE_WORDS * 2]`.